File: Android/livestreamcore/src/main/java/com/trtc/uikit/livekit/livestreamcore/manager/module/RoomManager.java
package com.trtc.uikit.livekit.livestreamcore.manager.module;

import static com.trtc.uikit.livekit.livestreamcore.state.RoomState.LiveStatus.NONE;
import static com.trtc.uikit.livekit.livestreamcore.state.RoomState.LiveStatus.PLAYING;
import static com.trtc.uikit.livekit.livestreamcore.state.RoomState.LiveStatus.PUSHING;

import android.content.Context;
import android.text.TextUtils;
import android.util.Log;

import com.tencent.cloud.tuikit.engine.common.TUICommonDefine;
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.trtc.tuikit.common.foregroundservice.VideoForegroundService;
import com.trtc.tuikit.common.system.ContextProvider;
import com.trtc.uikit.livekit.livestreamcore.R;
import com.trtc.uikit.livekit.livestreamcore.common.Constants;
import com.trtc.uikit.livekit.livestreamcore.common.utils.Logger;
import com.trtc.uikit.livekit.livestreamcore.manager.api.ILiveStream;
import com.trtc.uikit.livekit.livestreamcore.state.LiveStreamState;

import org.json.JSONException;
import org.json.JSONObject;

public class RoomManager extends BaseManager {
    private static final String TAG = "RoomManager";

    private RoomObserver mRoomObserver;

    public RoomManager(LiveStreamState state, ILiveStream service) {
        super(state, service);
    }

    public void setRoomObserver(RoomObserver observer) {
        mRoomObserver = observer;
    }

    @Override
    public void destroy() {
    }

    public void initCreateRoomState(String roomId, int maxCount) {
        Logger.info(TAG + " initCreateRoomState roomId [roomId: " + roomId + ", maxCount:" + maxCount + "]");
        mVideoLiveState.roomState.roomId = roomId;
        mVideoLiveState.roomState.maxCoGuestCount.set(maxCount);
        mVideoLiveState.roomState.ownerInfo.userId = mVideoLiveState.userState.selfInfo.userId;
        mVideoLiveState.roomState.ownerInfo.userName = mVideoLiveState.userState.selfInfo.userName;
        mVideoLiveState.roomState.ownerInfo.avatarUrl = mVideoLiveState.userState.selfInfo.avatarUrl;
        mVideoLiveState.userState.selfInfo.userRole = TUIRoomDefine.Role.ROOM_OWNER;
    }

    public void startLive(TUIRoomDefine.RoomInfo roomInfo, TUIRoomDefine.GetRoomInfoCallback callback) {
        Logger.info(TAG + " start[roomId" + mVideoLiveState.roomState.roomId
                + ",liveService:" + mVideoLiveService.hashCode() + "]");
        dataReport();
        if (TextUtils.isEmpty(mVideoLiveState.roomState.roomId)) {
            Logger.error(TAG + " not init create room state");
            if (callback != null) {
                callback.onError(TUICommonDefine.Error.INVALID_PARAMETER, "roomId is empty");
            }
            return;
        }

        mVideoLiveState.roomState.liveStatus.set(PUSHING);
        mVideoLiveService.createRoom(roomInfo, new TUIRoomDefine.ActionCallback() {
            @Override
            public void onSuccess() {
                mVideoLiveService.enterRoom(roomInfo.roomId, new TUIRoomDefine.GetRoomInfoCallback() {
                    @Override
                    public void onSuccess(TUIRoomDefine.RoomInfo roomInfo) {
                        startForegroundService();
                        updateRoomState(roomInfo);
                        if (callback != null) {
                            callback.onSuccess(roomInfo);
                        }
                    }

                    @Override
                    public void onError(TUICommonDefine.Error error, String message) {
                        mVideoLiveState.roomState.liveStatus.set(NONE);
                        if (callback != null) {
                            callback.onError(error, message);
                        }
                    }
                });
            }

            @Override
            public void onError(TUICommonDefine.Error error, String message) {
                mVideoLiveState.roomState.liveStatus.set(NONE);
                if (callback != null) {
                    callback.onError(error, message);
                }
            }
        });
    }

    public void joinLive(String roomId, TUIRoomDefine.GetRoomInfoCallback callback) {
        Logger.info(TAG + " join room [roomId: " + roomId + ",liveService:" + mVideoLiveService.hashCode() + "]");
        dataReport();
        if (TextUtils.isEmpty(roomId)) {
            callback.onError(TUICommonDefine.Error.INVALID_PARAMETER, "roomId is empty");
            return;
        }
        mVideoLiveService.enterRoom(roomId, new TUIRoomDefine.GetRoomInfoCallback() {
            @Override
            public void onSuccess(TUIRoomDefine.RoomInfo roomInfo) {
                startForegroundService();
                updateRoomState(roomInfo);
                mVideoLiveState.roomState.liveStatus.set(PLAYING);
                if (callback != null) {
                    callback.onSuccess(roomInfo);
                }
            }

            @Override
            public void onError(TUICommonDefine.Error error, String message) {
                if (callback != null) {
                    callback.onError(error, message);
                }
            }
        });
    }

    public void leaveLive(TUIRoomDefine.ActionCallback callback) {
        mVideoLiveService.exitRoom(new TUIRoomDefine.ActionCallback() {
            @Override
            public void onSuccess() {
                stopForegroundService();
                if (callback != null) {
                    callback.onSuccess();
                }
            }

            @Override
            public void onError(TUICommonDefine.Error error, String message) {
                if (callback != null) {
                    callback.onError(error, message);
                }
            }
        });
    }

    public void stopLive(TUIRoomDefine.ActionCallback callback) {
        Logger.info(TAG + " stop room [roomId: " + mVideoLiveState.roomState.roomId + "]");
        mVideoLiveService.destroyRoom(new TUIRoomDefine.ActionCallback() {
            @Override
            public void onSuccess() {
                stopForegroundService();
                if (callback != null) {
                    callback.onSuccess();
                }
            }

            @Override
            public void onError(TUICommonDefine.Error error, String message) {
                if (callback != null) {
                    callback.onError(error, message);
                }
            }
        });
    }

    public void updateRoomState(TUIRoomDefine.RoomInfo roomInfo) {
        mVideoLiveState.roomState.updateState(roomInfo);
    }

    private void notifyRoomDismissed(String roomId) {
        if (mRoomObserver != null) {
            mRoomObserver.onRoomDismissed(roomId);
        }
    }

    private void dataReport() {
        try {
            JSONObject params = new JSONObject();
            params.put("framework", Constants.DATA_REPORT_FRAMEWORK);
            params.put("component", Constants.DATA_REPORT_COMPONENT);
            params.put("language", Constants.DATA_REPORT_LANGUAGE_JAVA);

            JSONObject jsonObject = new JSONObject();
            jsonObject.put("api", "setFramework");
            jsonObject.put("params", params);
            mVideoLiveService.callExperimentalAPI(jsonObject.toString());
        } catch (JSONException e) {
            Log.e(TAG, "dataReport:", e);
        }
    }

    public void onRoomDismissed(String roomId) {
        mVideoLiveState.roomState.liveStatus.set(NONE);
        notifyRoomDismissed(roomId);
    }

    public interface RoomObserver {
        void onRoomDismissed(String roomId);
    }

    private void startForegroundService() {
        Logger.info(TAG + " startForegroundService");
        Context context = ContextProvider.getApplicationContext();
        VideoForegroundService.start(context,
                context.getString(context.getApplicationInfo().labelRes),
                context.getString(R.string.livestreamcore_app_running),
                0);
    }

    private void stopForegroundService() {
        Logger.info(TAG + " stopForegroundService");
        Context context = ContextProvider.getApplicationContext();
        VideoForegroundService.stop(context);
    }
}

File: iOS/LiveStreamCore/Sources/Manager/Module/RoomManager.swift
//
//  RoomManager.swift
//  LiveStreamCore
//
//  Created by jeremiawang on 2024/10/23.
//

import RTCCommon
import RTCRoomEngine

class RoomManager {
    let observerState = ObservableState<RoomState>(initialState: RoomState())
    var roomState: RoomState {
        observerState.state
    }
    
    private weak var context: LiveStreamManager.Context?
    private let service: LiveStreamService
    
    init(context: LiveStreamManager.Context) {
        self.context = context
        self.service = context.service
    }
    
    func initCreateRoomState(roomId: String, maxCount: Int) {
        guard let context = context else { return }
        observerState.update(isPublished: false) { roomState in
            roomState.roomId = roomId
            roomState.maxCoGuestCount = maxCount
            roomState.ownerInfo.userId = context.userManager.userState.selfInfo.userId
            roomState.ownerInfo.userName = context.userManager.userState.selfInfo.userName
            roomState.ownerInfo.avatarUrl = context.userManager.userState.selfInfo.avatarUrl
        }
    }
    
    func startLive(roomInfo: TUIRoomInfo) async throws -> TUIRoomInfo {
        if roomInfo.roomId.isEmpty {
            throw LiveStreamCoreError.error(code: TUIError.invalidParameter, message: "roomId is Empty")
        }
        roomInfo.roomType = .live
        roomInfo.isSeatEnabled = true
        roomInfo.seatMode = .applyToTake
        
        context?.coHostManager.setEnableConnection(enable: true)
        context?.coGuestManager.setEnableConnection(enable: true)
        initCreateRoomState(roomId: roomInfo.roomId, maxCount: roomInfo.maxSeatCount)
        dataReport()
        do {
            try await service.createRoom(roomInfo: roomInfo)
            let roomInfo = try await service.enterRoom(roomId: roomInfo.roomId)
            onStartLiveSuccess(roomInfo: roomInfo)
            return roomInfo
        } catch let LiveStreamCoreError.error(code, message) {
            LiveStreamLog.error("\(#file)","\(#line)","create:[roomId:\(roomInfo.roomId),code:\(code),message:\(message)]")
            onLeaveRoom()
            throw LiveStreamCoreError.error(code: code, message: message)
        }
    }
    
    func joinLive(roomId: String) async throws -> TUIRoomInfo {
        dataReport()
        do {
            let roomInfo = try await service.enterRoom(roomId: roomId)
            onJoinLiveSuccess(roomInfo: roomInfo)
            return roomInfo
        } catch let LiveStreamCoreError.error(code, message){
            LiveStreamLog.error("\(#file)","\(#line)","create:[roomId:\(roomId),code:\(code),message:\(message)]")
            onLeaveRoom()
            throw LiveStreamCoreError.error(code: code, message: message)
        }
    }
    
    func leaveLive() async throws {
        do {
            try await service.exitRoom()
            onLeaveRoom()
        } catch let LiveStreamCoreError.error(code, message){
            LiveStreamLog.error("\(#file)","\(#line)","leave:[code:\(code),message:\(message)]")
            throw LiveStreamCoreError.error(code: code, message: message)
        } catch {
            assert(false, "unknown error, description:[:\(error.localizedDescription)]")
        }
    }
    
    func stopLive() async throws {
        do {
            try await service.destroyRoom()
            onLeaveRoom()
        } catch let LiveStreamCoreError.error(code, message){
            LiveStreamLog.error("\(#file)","\(#line)","leave:[code:\(code),message:\(message)]")
            throw LiveStreamCoreError.error(code: code, message: message)
        } catch {
            assert(false, "unknown error, description:[:\(error.localizedDescription)]")
        }
    }
    
    func updateOwnerInfo(roomInfo: TUIRoomInfo) {
        observerState.update(isPublished: false) { roomState in
            roomState.ownerInfo.userId = roomInfo.ownerId
            roomState.ownerInfo.userName = roomInfo.ownerName
            roomState.ownerInfo.avatarUrl = roomInfo.ownerAvatarUrl
        }
    }
}

// MARK: - Observer
extension RoomManager {
    func onRoomDismissed(roomId: String) {
        onLeaveRoom()
    }
}

// MARK: - Private
extension RoomManager {
    private func dataReport() {
        do {
            let apiParams: [String : Any] = [
                "api": "setFramework",
                "params": [
                    "framework": Constants.DataReport.framework,
                    "component": Constants.DataReport.component,
                    "language": Constants.DataReport.language,
                ],
              ]
            let jsonData = try JSONSerialization.data(withJSONObject: apiParams, options: .prettyPrinted)
            if let jsonString = String(data: jsonData, encoding: .utf8) {
                service.callExperimentalAPI(jsonStr: jsonString)
            } else {
                print("Error converting JSON data to string")
            }
        } catch {
            print("Error converting dictionary to JSON: \(error.localizedDescription)")
        }
    }
    
    private func updateRoomState(roomInfo: TUIRoomInfo) {
        observerState.update(isPublished: false) { roomState in
            roomState.roomId = roomInfo.roomId
            roomState.ownerInfo.userId = roomInfo.ownerId
            roomState.maxCoGuestCount = roomInfo.maxSeatCount
        }
    }
    
    private func onStartLiveSuccess(roomInfo: TUIRoomInfo) {
        modifyRoomState(value: .pushing, keyPath: \RoomState.liveStatus, isPublished: true)
        context?.userManager.onStartLiveSuccess()
        updateRoomState(roomInfo: roomInfo)
        updateOwnerInfo(roomInfo: roomInfo)
        Task {
            try? await context?.coGuestManager.initConnectedGuestList()
        }
        Task {
            try? await context?.coGuestManager.initGuestApplicationList()
        }
    }
    
    private func onJoinLiveSuccess(roomInfo: TUIRoomInfo) {
        updateRoomState(roomInfo: roomInfo)
        modifyRoomState(value: .playing, keyPath: \RoomState.liveStatus, isPublished: true)
        if context?.userManager.userState.selfInfo.userId == roomInfo.ownerId {
            context?.userManager.onStartLiveSuccess()
        }
        updateOwnerInfo(roomInfo: roomInfo)
        Task {
            try? await context?.coGuestManager.initConnectedGuestList()
        }
    }
    
    private func onLeaveRoom() {
        modifyRoomState(value: .none, keyPath: \RoomState.liveStatus, isPublished: true)
        modifyRoomState(value: TUIUserInfo(), keyPath: \RoomState.ownerInfo)
        modifyRoomState(value: 0, keyPath: \RoomState.maxCoGuestCount)
        context?.coHostManager.onLeaveRoom()
        context?.coGuestManager.onLeaveRoom()
        context?.mediaManager.onLeaveRoom()
        context?.userManager.onLeaveRoom()
        context?.viewManager.onLeaveRoom()
    }
    
    private func modifyRoomState<T>(value: T, keyPath: WritableKeyPath<RoomState, T>, isPublished: Bool = false) {
        observerState.update(isPublished: isPublished) { roomState in
            roomState[keyPath: keyPath] = value
        }
    }
}
